File: tools/test_map_render.mjs
import assert from 'node:assert/strict';
import {readFile,writeFile,mkdir} from 'node:fs/promises';
import createEngine from './map_editor/generated/engine.mjs';
import {renderOverview,renderFullPng} from './map_editor/render.mjs';
const root=new URL('../',import.meta.url),folder=new URL('build/map-recipe-tests/',root);
const cases=JSON.parse(await readFile(new URL('render-manifest.json',folder),'utf8'));
const art=JSON.parse(await readFile(new URL('tools/map_editor/generated/art.json',root),'utf8'));
const engine=await createEngine();
await mkdir(new URL('artifacts/map_editor/',root),{recursive:true});
function run(program,seed){engine.HEAPU32.set(new Uint32Array(program.flat()),engine._recipe_input()>>>2);assert.equal(engine._recipe_run(program.length,seed),0);}
for(const f of cases){
  run(f.world,f.seed);
  if(f.materials.length){run(f.materials,f.seed);assert.equal(engine._recipe_apply_materials(),0);}
  const native=await readFile(new URL(`${f.name}.render`,folder));
  const ground=engine._recipe_ground(),ptr=engine._recipe_tiles(),tileMap=engine.HEAPU16.slice(ptr>>>1,(ptr>>>1)+1024*1024);
  assert.deepEqual(Buffer.from(engine.HEAPU8.slice(ground,ground+4096)),native.subarray(4176,8272),`${f.name} ground IDs`);
  assert.deepEqual(Buffer.from(tileMap.buffer),native.subarray(8272,-4),`${f.name} every rendered tile`);
  assert.equal(engine._recipe_render_signature()>>>0,native.readUInt32LE(native.length-4),`${f.name} tiles and grip`);
  if(['original','natural','roads','roads-wide','active'].includes(f.name)){
    const png=await renderFullPng(tileMap,art);
    await writeFile(new URL(`artifacts/map_editor/${f.name}-full-map.png`,root),Buffer.from(await png.arrayBuffer()));
    const preview=renderOverview(tileMap,art);
    await writeFile(new URL(`${f.name}-overview.rgba`,folder),preview.pixels);
  }
}
await writeFile(new URL('artifacts/map_editor/render-results.json',root),JSON.stringify({cases:cases.length,tilesPerCase:1048576,nativeWasmTileParity:true,groundIdParity:true,surfaceParity:true,fullPngSize:[8192,8192]},null,2)+'\n');
console.log(`PASS ${cases.length} native/WASM full tile maps, ground IDs and surface signatures; exported original and natural 8192-square PNGs.`);

File: tools/test_map_render.py
"""Check PNG encoding, source tile pixels, and the downsampled editor preview."""
import json
import struct
from pathlib import Path
from PIL import Image, ImageChops

ROOT=Path(__file__).resolve().parents[1]
art=json.loads((ROOT/'tools/map_editor/generated/art.json').read_text())
for name in ('original','natural','roads','roads-wide','active'):
    path=ROOT/f'artifacts/map_editor/{name}-full-map.png'
    with Image.open(path) as check:
        check.verify()
    with Image.open(path) as image:
        assert image.size==(8192,8192) and image.mode=='P'
        raw=(ROOT/f'build/map-recipe-tests/{name}.render').read_bytes()
        # Sample inside tiles, at borders, and throughout all map quadrants.
        for y in range(0,8192,37):
            for x in range(0,8192,43):
                reference=struct.unpack_from('<H',raw,8272+((y//8)*1024+x//8)*2)[0]
                expected=art['tiles'][art['refs'][reference]*64+(y%8)*8+x%8]
                assert image.getpixel((x,y))==expected,(name,x,y)
        preview=Image.frombytes('RGBA',(2048,2048),(ROOT/f'build/map-recipe-tests/{name}-overview.rgba').read_bytes()).convert('RGB')
        reference=image.convert('RGB').resize((2048,2048),Image.Resampling.BOX)
        diff=ImageChops.difference(preview,reference)
        assert max(high for low,high in diff.getextrema())<=1
        preview.save(ROOT/f'artifacts/map_editor/{name}-overview.png')
        image.crop((3840,3840,4480,4480)).save(ROOT/f'artifacts/map_editor/{name}-detail.png')
print('PASS full PNG integrity, sampled source-art pixels, and overview downsampling.')
